core/execution/future/market.go

```diff
@@ -3986,5 +4308,113 @@ func (m *Market) GetRiskFactors() *types.RiskFactor {
 }
 
 func (m *Market) UpdateMarginMode(ctx context.Context, party string, marginMode types.MarginMode, marginFactor num.Decimal) error {
-	return errors.New("Unsupported")
+	return m.switchMarginMode(ctx, party, marginMode, marginFactor)
+}
+
+func (m *Market) getMarginMode(party string) types.MarginMode {
+	marginFactor, ok := m.partyMarginFactor[party]
+	if !ok || marginFactor.IsZero() {
+		return types.MarginModeCrossMargin
+	}
+	return types.MarginModeIsolatedMargin
+}
+
+func (m *Market) useGeneralAccountForMarginSearch(party string) bool {
+	return m.getMarginMode(party) == types.MarginModeCrossMargin
+}
+
+func (m *Market) getMarginFactor(party string) num.Decimal {
+	marginFactor, ok := m.partyMarginFactor[party]
+	if !ok || marginFactor.IsZero() {
+		return num.DecimalZero()
+	}
+	return marginFactor
+}
+
+// switchMarginMode handles a switch between margin modes and/or changes to the margin factor.
+// When switching to isolated margin mode, the following steps will be taken:
+// 1. For any active position, calculate average entry price * abs(position) * margin factor.
+// Calculate the amount of funds which will be added to, or subtracted from, the general account in order to do this.
+// If additional funds must be added which are not available, reject the transaction immediately.
+// 2. For any active orders, calculate the quantity limit price * remaining size * margin factor which needs to be placed in the
+// order margin account. Add this amount to the difference calculated in step 1. If this amount is less than or equal to the
+// amount in the general account, perform the transfers (first move funds into/out of margin account, then move funds into
+// the order margin account). If there are insufficient funds, reject the transaction.
+// 3. Move account to isolated margin mode on this market
+//
+// When switching from isolated margin mode to cross margin mode, the following steps will be taken:
+// 1. Any funds in the order margin account will be moved to the margin account.
+// 2. At this point trading can continue with the account switched to the cross margining account type.
+// If there are excess funds in the margin account they will be freed at the next margin release cycle.
+func (m *Market) switchMarginMode(ctx context.Context, party string, marginMode types.MarginMode, marginFactor num.Decimal) error {
```

Comment:
would be useful for the frontend here to add some sort of event, indicating that the party is changing marging mode.

Also to store in the datanode: the party, margin mode, and eventual factor. that'll be useful for UIs to know what's the user selected for margin mode. I realise that it's also set in the margin level, but if the user has never traded, the frontend won't be able to know using that.

Reply:
I don't understand. it's already there, in margin levels.

Reply:
it's there in margin level, but in the case the party do not have a position or open orders yet, or there's no market observable etc, then it's never sent to the data node and there's no way to expose that to the API. for the frontends we'll need most likely something exposing: market_id, party_id, margin_mode, margin_factor.

We'll also need to send an event for that on the protocol upgrade to default this to cross most likely.